rasa/shared/core/domain.py

```diff
@@ -252,195 +255,213 @@ def from_directory(cls, path: Text) -> "Domain":
+            )
+        return combined, duplicates
 
+    @staticmethod
     def merge_domain_dicts(
-        self, domain1: Dict, domain2: Dict, override: bool = False
-    ) -> Dict[Text, Any]:
-        """Merges this domain dict with another one, combining their attributes.
+        domain_dict: Dict, combined: Dict, override: bool = False, is_dir: bool = False
+    ) -> Dict:
+        """Combines two domain dictionaries."""
+        if not domain_dict:
+            return combined
 
-        This is used when multiple domain yml files are configured in a single
-        directory. Unlike the merge method above, which merges Domain objects by
-        creating each object then merging it with the previous, this method merges
-        domain dicts, and ensures all attributes (like ``intents``, ``entities``, and
-        ``actions``) are known to the Domain when the object is created.
+        if not combined:
+            return domain_dict
 
-        List attributes like ``intents`` and ``actions`` are deduped
-        and merged. Single attributes are taken from `domain1` unless
-        override is `True`, in which case they are taken from `domain2`.
-        """
-        if not domain2:
-            return domain1
-
-        if not domain1:
-            return domain2
-
-        domain_dict = domain2
-        combined = domain1
+        if not combined.get(KEY_TRAINING_DATA_FORMAT_VERSION):
+            combined[
+                KEY_TRAINING_DATA_FORMAT_VERSION
+            ] = LATEST_TRAINING_DATA_FORMAT_VERSION
 
         if override:
             config = domain_dict["config"]
             for key, val in config.items():
                 combined["config"][key] = val
 
-        if override or domain2.get("session_config"):
+        if (
+            override
+            or (
+                not is_dir
+                and combined.get(SESSION_CONFIG_KEY)
+                == SessionConfig.default().as_dict()
+            )
+            or (is_dir and domain_dict.get(SESSION_CONFIG_KEY))
+        ):
             combined[SESSION_CONFIG_KEY] = domain_dict[SESSION_CONFIG_KEY]
 
         duplicates: Dict[Text, List[Text]] = {}
 
-        for key in [KEY_INTENTS, KEY_ENTITIES]:
-            if combined.get(key) or domain_dict.get(key):
-                duplicates[key] = self.extract_duplicates(
-                    combined.get(key, []), domain_dict.get(key, [])
-                )
-                combined[key] = combined.get(key, [])
-                domain_dict[key] = domain_dict.get(key, [])
-                combined[key] = self.merge_lists_of_dicts(
-                    combined[key], domain_dict[key], override
-                )
+        combined, duplicates = Domain._merge_intents_and_entities(
+            combined, domain_dict, duplicates, override
+        )
 
         # remove existing forms from new actions
         for form in combined.get(KEY_FORMS, []):
```

Comment:
Do we still need to do this?

Reply:
Since this was part of the original code, I'd say so - the duplicate check is only between the same keys, it won't check for duplicate form names in the other domain's actions.